sync-core/src/main/java/com/cloudant/http/HttpConnection.java

```diff
@@ -0,0 +1,234 @@
+        this.inputLength = input.length;
+    }
+
+    /**
+     * Set the InputStream of data to be sent to the server
+     * @param input InputStream of data to be sent to the server
+     */
+    public void setInput(InputStream input) {
+        this.input = input;
+    }
+
+    /**
+     * Set the InputStream of data, of known length, to be sent to the server
+     * @param input InputStream of data to be sent to the server
+     * @param inputLength Length of data to be sent to the server, in bytes
+     */
+    public void setInput(InputStream input, long inputLength) {
+        this.input = input;
+        this.inputLength = inputLength;
+    }
+
+    private void executeInternal() throws IOException {
+        connection = (HttpURLConnection) url.openConnection();
+        for (String key : requestProperties.keySet()) {
+            connection.setRequestProperty(key, requestProperties.get(key));
+        }
+        // always read the result, so we can retrieve the HTTP response code
+        connection.setDoInput(true);
+        connection.setRequestMethod(requestMethod);
+        if (contentType != null) {
+            connection.setRequestProperty("Content-type", contentType);
+        }
+        if (username != null) {
+            String encodedAuth = Base64.encodeBase64String(
+                    String.format("%s:%s", username, password != null ? password : "")
+                            .getBytes());
+            connection.setRequestProperty("Authorization", String.format("Basic %s", encodedAuth));
+        }
+        if (input != null) {
+            connection.setDoOutput(true);
+            if (inputLength != -1) {
+                // TODO on 1.7 upwards this method takes a long, otherwise int
+                connection.setFixedLengthStreamingMode((int)this.inputLength);
+            } else {
+                // TODO some situations where we can't do chunking, like multipart/related
+                /// https://issues.apache.org/jira/browse/COUCHDB-1403
+                connection.setChunkedStreamingMode(1024);
+            }
+            connection.setRequestProperty("Expect", "100-continue");
+
+            // this will throw an exception if it doesn't get a 100-continue
+            int bufSize = 1024;
+            int nRead = 0;
+            byte[] buf = new byte[bufSize];
+            InputStream is = input;
+            OutputStream os = connection.getOutputStream();
+
+            while((nRead = is.read(buf)) >= 0) {
+                os.write(buf, 0, nRead);
+            }
+        }
+    }
+
+    /**
+     * Execute request without returning data from server.
+     * @throws IOException
+     */
+    public void execute() throws IOException {
+        executeInternal();
+    }
+
+    /**
+     * Execute request and return data from server as a String
+     * @return String of data from server, if any
+     * @throws IOException
+     */
+    public String executeToString() throws IOException {
```

Comment:
`responseAsString()` would be a lie because no I/O gets done until this call is made. It's really "Execute the request and get the response as a string". If we're happy to lie a little bit for the sake of brevity, then `responseAsString()` is acceptable.